### PR TITLE
Add HS4xD support

### PR DESCRIPTION
**Pull request by kokas-a**
Minimal HSDK4xD support

### CODE REVIEW

**Comment by kokas-a** [comment]
@evgeniy-paltsev Could you please  take a look? PR has same issues with checks, but it wold be nice to have your feedback on key points

File: arch/arc/core/reset.S [hunk 1]
```diff
@@ -86,6 +86,24 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_and__start,__start)
...
 #endif
 
+#if defined (CONFIG_CPU_HS4X) && defined (CONFIG_ARC_HAS_DSP)
+	/* case of HS4XD cpu */
+	lr	 r0, [_ARC_V2_LPB_BUILD]
+	breq r0, 0, skip_lpb_setup
+	mov  r0, 1	/* disable LPB for HS4XD */
+	sr   r0, [_ARC_V2_LPB_CTRL]
+skip_lpb_setup:
+#endif
```
Comment (evgeniy-paltsev) [inline arch/arc/core/reset.S:96]: Could you please check if there is simple way to move it to SoC code? Preferably with just using ASM macro.

Another thing - we need to move this after cache maitanance code in reset.S

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]

File: arch/arc/Kconfig [hunk 2]
```diff
@@ -278,13 +286,18 @@ config CODE_DENSITY
...
+	default y if CPU_HS3X || CPU_HS4X || CPU_HS5X || CPU_HS6X
 	help
 	  Depending on the configuration, CPU can contain accumulator reg-pair
 	  (also referred to as r58:r59). These can also be used by gcc as GPR so
 	  kernel needs to save/restore per process
 
+config ARC_HAS_DSP
+	bool "Dsp support"
+	help
+		Some CPUs contain HW DSP extension
```
Comment (evgeniy-paltsev) [inline arch/arc/Kconfig:299]: Indent issue

File: arch/arc/Kconfig [hunk 3]
```diff
@@ -278,13 +286,18 @@ config CODE_DENSITY
...
-	bool "Reg Pair ACCL:ACCH (FPU and/or MPY > 6)"
-	default y if CPU_HS3X || CPU_HS5X || CPU_HS6X
+	bool "Reg Pair ACCL:ACCH (FPU and/or MPY > 6 and/or DSP)"
+	default y if CPU_HS3X || CPU_HS4X || CPU_HS5X || CPU_HS6X
 	help
 	  Depending on the configuration, CPU can contain accumulator reg-pair
 	  (also referred to as r58:r59). These can also be used by gcc as GPR so
 	  kernel needs to save/restore per process
 
+config ARC_HAS_DSP
```
Comment (evgeniy-paltsev) [inline arch/arc/Kconfig:296]: We already have ARC_HAS_DSP Kconfig symbol:

https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arc/core/dsp/Kconfig#L6

File: arch/arc/core/reset.S [hunk 4]
```diff
@@ -86,6 +86,24 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_and__start,__start)
...
+
+#ifdef CONFIG_ARC_HAS_DSP
+	lr	 r0, [_ARC_V2_DSP_BUILD]
+	bmsk r0, r0, 7
+	breq r0, 0, skip_dsp_setup
+	mov	 r0, 0 /* DSP_CTRL_DISABLED_ALL */
+	sr	 r0, [_ARC_V2_DSP_CTRL]
+skip_dsp_setup:
+#endif /* CONFIG_ARC_HAS_DSP */
```
Comment (evgeniy-paltsev) [inline arch/arc/core/reset.S:105]: We should move it to macro to https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arc/core/dsp/

And this seems to interfere with existing DSP functionality.

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]

File: boards/arc/hsdk4xd/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+#
+# Copyright (c) 2022 Synopsys, Inc. All rights reserved.
```
Comment (evgeniy-paltsev) [inline boards/arc/hsdk4xd/CMakeLists.txt:2]: Minor issue: here and in other plases we use 2022 instead of 2023

File: soc/arc/snps_arc_hsdk4xd/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library_include_directories(${ZEPHYR_BASE}/drivers)
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hsdk4xd/CMakeLists.txt:3]: Hmm, I'm wondering if it's actually required.

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]

File: arch/arc/core/reset.S [hunk 7]
```diff
@@ -16,6 +16,9 @@
 #include <zephyr/arch/cpu.h>
 #include <swap_macros.h>
 #include <zephyr/arch/arc/asm-compat/assembler.h>
+#ifdef CONFIG_SOC_ARC_HSDK4XD
```
Comment (evgeniy-paltsev) [inline arch/arc/core/reset.S:19]: I'd make this more generic. I.e. by not using exact SoC Kconfig and by replacing it with "CONFIG_EARLY_SOC_INIT" or something like that.

File: arch/arc/core/reset.S [hunk 8]
```diff
@@ -109,6 +112,12 @@ done_icache_invalidate:
 
 done_dcache_invalidate:
 
+#ifdef CONFIG_SOC_ARC_HSDK4XD
+	_soc_early_tune
```
Comment (evgeniy-paltsev) [inline arch/arc/core/reset.S:116]: I'd probably even make it more detailed. I.e.: `soc_early_init_percpu` or `soc_early_asm_init_percpu`

File: boards/arc/hsdk4xd/support/openocd.cfg [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+set _coreid [expr {$_coreid + 1}]
+set _dbgbase [expr {0x00000000 | ($_coreid << 13)}]
+arc_hs_init_regs
+
+# Enable L2 cache support for core 1.
+$_TARGETNAME1 arc cache l2 auto 1
+
+target smp $_TARGETNAME1 $_TARGETNAME2 $_TARGETNAME3 $_TARGETNAME4
+
+# vi:ft=tcl
```
Comment (evgeniy-paltsev) [inline boards/arc/hsdk4xd/support/openocd.cfg:116]: Probably not needed.

File: arch/arc/core/reset.S (hunk 7 above)
Comment (kokas-a) [reply arch/arc/core/reset.S:19]: In generic way this will require to create "soc_ctrl.h" header for each SOC. Is it fine?

Comment (evgeniy-paltsev) [reply arch/arc/core/reset.S]: We still can guard this include under ifdef. But we can do ifdef against generic Kconfig symbol instead of SoC specific one.

Comment (kokas-a) [reply arch/arc/core/reset.S]: please check the recent changes

File: boards/arc/hsdk4xd/CMakeLists.txt (hunk 5 above)
Comment (evgeniy-paltsev) [reply boards/arc/hsdk4xd/CMakeLists.txt:2]: This is marked as resolved, though 2022 year is still used in many places here.

File: arch/arc/Kconfig [hunk 10]
```diff
@@ -377,6 +385,13 @@ config ARC_EXCEPTION_DEBUG
 	  and parameters, at a cost of code/data size for the human-readable
 	  strings.
 
+config ARC_EARLY_SOC_INIT
+	bool "Make early satge SoC-specific initialization"
+	default n
```
Comment (evgeniy-paltsev) [inline arch/arc/Kconfig:390]: We can drop `default n` - it's n by default

File: arch/arc/Kconfig [hunk 11]
```diff
@@ -89,6 +89,14 @@ config CPU_HS3X
 	help
 	  If y, the SoC uses an ARC HS3x or HS4x CPU
```
Comment (evgeniy-paltsev) [inline arch/arc/Kconfig:90]: We should change this - otherwise it's confusing a bit

File: arch/arc/Kconfig [hunk 12]
```diff
@@ -89,6 +89,14 @@ config CPU_HS3X
...
 
+config CPU_HS4X
```
Comment (evgeniy-paltsev) [inline arch/arc/Kconfig:92]: We should add default mcpu value for hs4x as we are adding new cpu type.

Otherwise mcpu would be empty (if we don't use mcpu overwrite feature).

IOW: It would work for this board, but we should provide reasonable default (mcpu=hs4x)

File: boards/arc/hsdk4xd/CMakeLists.txt (hunk 5 above)
Comment (kokas-a) [reply boards/arc/hsdk4xd/CMakeLists.txt:2]: From previous comment I understood it like "We SHOULD use 22 instead of 23". So I turned all dates back to 2022 =)

File: arch/arc/Kconfig (hunk 12 above)
Comment (kokas-a) [reply arch/arc/Kconfig:92]: Done. Ref. cmake/gcc-m-cpu.cmake

File: arch/arc/core/dsp/Kconfig [hunk 13]
```diff
@@ -16,6 +16,11 @@ config ARC_DSP
 	help
 	  This option enables DSP and DSP instructions.
 
+config ARC_DSP_TURNED_OFF
+	bool "Turn off DSP if it presents"
```
Comment (evgeniy-paltsev) [inline arch/arc/core/dsp/Kconfig:20]:
```suggestion
	bool "Turn off DSP if it presents"
	depends on !ARC_DSP
```

File: arch/arc/core/dsp/swap_dsp_macros.h [hunk 14]
```diff
@@ -260,3 +260,14 @@ dsp_skip_load :
...
+.macro _dsp_extension_probe
+	lr	 r0, [_ARC_V2_DSP_BUILD]
+	bmsk r0, r0, 7
+	breq r0, 0, skip_dsp_setup
+#ifdef CONFIG_ARC_DSP_TURNED_OFF
+	mov	 r0, 0 /* DSP_CTRL_DISABLED_ALL */
+	sr	 r0, [_ARC_V2_DSP_CTRL]
+#endif
+skip_dsp_setup :
+.endm
```
Comment (evgeniy-paltsev) [inline arch/arc/core/dsp/swap_dsp_macros.h:273]:
```suggestion
.macro _dsp_extension_probe
#ifdef CONFIG_ARC_DSP_TURNED_OFF
	lr	 r0, [_ARC_V2_DSP_BUILD]
	bmsk r0, r0, 7
	breq r0, 0, skip_dsp_setup
	mov	 r0, 0 /* DSP_CTRL_DISABLED_ALL */
	sr	 r0, [_ARC_V2_DSP_CTRL]
skip_dsp_setup :
#endif
.endm
```

File: soc/arc/snps_arc_hsdk4xd/soc_ctrl.h [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+
+#ifndef _ARC_HSDK4XD_SOC_CTRL_H_
+#define _ARC_HSDK4XD_SOC_CTRL_H_
+
+#ifdef _ASMLANGUAGE
+.macro soc_early_asm_init_percpu
+	/* case of HS4XD cpu */
+	lr	 r0, [_ARC_V2_LPB_BUILD]
+	breq r0, 0, skip_lpb_setup
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hsdk4xd/soc_ctrl.h:15]: As this is code for one exact SoC which is ASIC (so it won't be changed as FPGA based SoCs) - we can drop BCR check here.

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]

File: arch/arc/core/dsp/swap_dsp_macros.h (hunk 14 above)
Comment (evgeniy-paltsev) [reply arch/arc/core/dsp/swap_dsp_macros.h:273]: Probably even drop BCR check - as we assume that platform should set `CONFIG_ARC_HAS_DSP` proper way.

File: dts/arc/synopsys/arc_hs4xd.dtsi [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+			reg = <0xf0003000 0x80>;
+			ngpios = <24>;				/* FIXME */
+			interrupt-parent = <&idu_intc>;
+
+			gpio-controller;
+			#gpio-cells = <2>;
+
+			status = "disabled";
+		};
+/* FIXME */
```
Comment (evgeniy-paltsev) [inline dts/arc/synopsys/arc_hs4xd.dtsi:133]: I'm wondering if this is still related to some issue?

File: soc/arc/snps_arc_hsdk4xd/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+
+if(COMPILER STREQUAL gcc)
+  # GNU compiler options
+  zephyr_compile_options(-mno-sdata)
+  zephyr_compile_options_ifdef(CONFIG_FPU -mfpu=fpud_all)
+else()
+  # MWDT compiler options
+   zephyr_compile_options(-arcv2hs -core4 -Xdual_issue -Xcode_density -Hrgf_banked_regs=32 -HL
+    -Xatomic -Xll64 -Xunaligned -Xdiv_rem=radix4 -Xswap -Xbitscan -Xmpy_option=qmpyh -Xshift_assist
+    -Xbarrel_shifter -Xtimer0 -Xtimer1 -Xrtc -Hld_cycles=2 -Xlpb_size=128)
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hsdk4xd/CMakeLists.txt:11]:
```suggestion
  zephyr_compile_options(-arcv2hs -core4 -Xdual_issue -Xcode_density -Hrgf_banked_regs=32 -HL
    -Xatomic -Xll64 -Xunaligned -Xdiv_rem=radix4 -Xswap -Xbitscan -Xmpy_option=qmpyh -Xshift_assist
    -Xbarrel_shifter -Xtimer0 -Xtimer1 -Xrtc -Hld_cycles=2)
```

Alignment fix and drop Xlpb_size as we disable LPB anyway.

File: soc/arc/snps_arc_hsdk4xd/linker.ld [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+ * @brief Linker script for the HSDK4XD Development Kit
+ */
+
+#include <zephyr/devicetree.h>
+
+/*
+ * SRAM base address and size
+ */
+#if DT_NODE_HAS_PROP(DT_CHOSEN(zephyr_sram), reg) && \
+    (DT_REG_SIZE(DT_CHOSEN(zephyr_sram)) > 0)
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hsdk4xd/linker.ld:17]:
```suggestion
#if DT_NODE_HAS_PROP(DT_CHOSEN(zephyr_sram), reg) && (DT_REG_SIZE(DT_CHOSEN(zephyr_sram)) > 0)
```

This fits to 100 character per line limit.

File: arch/arc/Kconfig [hunk 19]
```diff
@@ -377,6 +385,12 @@ config ARC_EXCEPTION_DEBUG
 	  and parameters, at a cost of code/data size for the human-readable
 	  strings.
 
+config ARC_EARLY_SOC_INIT
+	bool "Make early satge SoC-specific initialization"
+	help
+	  Some SoCs required to be fine-tuned on very early stages of
+	  initialization.
```
Comment (evgeniy-paltsev) [inline arch/arc/Kconfig:392]: I'd rephrase it to:

```
Call SoC per-core setup code on early stage initialization (before C runtime initialization).
Setup code is called in form of soc_early_asm_init_percpu assembler macro.
```

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]

File: cmake/gcc-m-cpu.cmake [hunk 20]
```diff
@@ -89,6 +89,8 @@ elseif("${ARCH}" STREQUAL "arc")
     set(GCC_M_CPU em4_fpuda)
   elseif(CONFIG_CPU_HS3X)
     set(GCC_M_CPU archs)
+  elseif(CONFIG_CPU_HS4X)
+    set(GCC_M_CPU hs4x)
```
Comment (nordicjm) [inline cmake/gcc-m-cpu.cmake:93]: Can you add this to https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/mcumgr/grp/os_mgmt/include/os_mgmt_processor.h as a separate commit?

Comment (kokas-a) [reply cmake/gcc-m-cpu.cmake:93]: Done: https://github.com/zephyrproject-rtos/zephyr/pull/57348

File: boards/arc/hsdk4xd/Kconfig.defconfig [hunk 21]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2023 Synopsys, Inc. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_HSDK4XD
+
+config BOARD
+	default "hsdk4xd"
+
+# TODO
```
Comment (abrodkin) [inline boards/arc/hsdk4xd/Kconfig.defconfig:9]: Do we really need that empty `# TODO` tag here?

**Review by abrodkin — CHANGES_REQUESTED** [review]
A couple of minor notes below. And what about documentation update or maybe even a separate docs page for that board?

File: boards/arc/hsdk4xd/Kconfig.defconfig (hunk 21 above)
Comment (kokas-a) [reply boards/arc/hsdk4xd/Kconfig.defconfig:9]: Support of HSDK4xD in minimal configuration announced in this PR. Kconfig supposed to be extended with additional board features. ``` TODO ```  hints about that =)

Comment (abrodkin) [reply boards/arc/hsdk4xd/Kconfig.defconfig]: That's good, but still it worth saying what need to be done, as in 2 weeks we'll all forget about that.

Comment (ruuddw) [reply boards/arc/hsdk4xd/Kconfig.defconfig]: or even better, instead of this 'TODO' in the code, create a Github issue for it.

File: boards/arc/hsdk4xd/Kconfig.board [hunk 22]
```diff
@@ -0,0 +1,10 @@
+# DesignWare ARC HSDK4XD Development Kit board configuration
+
+# Copyright (c) 2023 Synopsys, Inc. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_HSDK4XD
+	bool "ARC HSDK4XD Development Kit"
+	depends on SOC_ARC_HSDK4XD
+	help
+	  TODO
```
Comment (evgeniy-paltsev) [inline boards/arc/hsdk4xd/Kconfig.board:10]: This is also shouldn't be just a TODO.

File: boards/arc/hsdk4xd/Kconfig.defconfig (hunk 21 above)
Comment (kokas-a) [reply boards/arc/hsdk4xd/Kconfig.defconfig]: https://github.com/zephyrproject-rtos/zephyr/issues/57430

**Review by abrodkin — APPROVED** [review]

File: arch/arc/core/reset.S [hunk 23]
```diff
@@ -109,6 +112,12 @@ done_icache_invalidate:
 
 done_dcache_invalidate:
 
+#ifdef CONFIG_ARC_EARLY_SOC_INIT
+	soc_early_asm_init_percpu
+#endif
+
+	_dsp_extension_probe
```
[the remark this answers is not in the document]
Comment (kokas-a) [reply arch/arc/core/reset.S:119]: Initially the macro looked like:
``` 
.macro _dsp_extension_probe
	lr	 r0, [_ARC_V2_DSP_BUILD]
	bmsk r0, r0, 7
	breq r0, 0, skip_dsp_setup
#ifdef CONFIG_ARC_DSP_TURNED_OFF
	mov	 r0, 0 /* DSP_CTRL_DISABLED_ALL */
	sr	 r0, [_ARC_V2_DSP_CTRL]
#endif
skip_dsp_setup :
.endm
```
After all stages of review it turned into what you may see.  Indeed, in current state it's more appropriate to move it into early_soc_init.

**Review by evgeniy-paltsev — APPROVED** [review]

File: arch/arc/core/reset.S (hunk 23 above)
[the remark this answers is not in the document]
Comment (ruuddw) [reply arch/arc/core/reset.S:119]: I see it was already merged. Please do so in a follow-up PR (this was minimal enablement only, anyway, and documentation is missing).
Does writing 0 to DSP_CTRL really 'turn DSP off"? Or just initializes some control flags?

[the remark this answers is not in the document]
Comment (kokas-a) [reply arch/arc/core/reset.S:119]: DSP_CTRL  does not really turn off DSP. Writing 0 disables some DSP features

[the remark this answers is not in the document]
Comment (ruuddw) [reply arch/arc/core/reset.S:119]: Thanks for confirming, that's what I remembered. So, not only the 'probe' but also the "DSP_TURNED_OFF" should be renamed. Actually, I don't think we need a kconfig option for that, it is just standard initialization.
